### PR TITLE
gguf : deduplicate

### PR DESCRIPTION
**Pull request by ggerganov**
Following things work:

- [x] `mmap`
- [x] `mlock`
- [x] 7B model:
  - [x] quantize
  - [x] main MacOS CPU
  - [x] main MacOS Metal
  - [x] main Ubuntu CPU
  - [x] main Ubuntu CUDA
  - [x] perplexity MacOS CPU
  - [x] perplexity MacOS Metal
  - [x] perplexity Ubuntu CPU
  - [x] perplexity Ubuntu CUDA
  - [x] simple
- [x] llama load/save session
- [ ] non-7B models (not tested)

### CODE REVIEW

**Comment by monatis** [comment]
@ggerganov  Btw, I removed the inner part of `llama_load_session_file_internal()` and its save counterpart, thinking of migrating them to GGUF format later on. So these need to be copied from old `llama.cpp` again if we don't want to do so. 

My idea was to use GGUF for all serialization / deserialization ops, but we may choose to keep things simple.

**Comment by klosax** [comment]
> Overall, it is looking good - need to test and fix CUDA

Compiling with cublas seems to work fine. Tested main perplexity quantize on openllama-3b-v2

**Comment by klosax** [comment]
Still got this compiler warning:
```
In file included from /usr/include/string.h:495,
                 from /usr/include/c++/9/cstring:42,
                 from llama.cpp:72:
In function ‘char* strncpy(char*, const char*, size_t)’,
    inlined from ‘int llama_token_to_str_with_model(const llama_model*, llama_token, char*, int)’ at llama.cpp:4793:20:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:34: warning: ‘char* __builtin_strncpy(char*, const char*, long unsigned int)’ output truncated before terminating nul copying 3 bytes from a string of the same length [-Wstringop-truncation]
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by klosax** [comment]
Maybe the warning above is the problem?

**Comment by klosax** [comment]
Line feed working now, I guess it was a typo `+` instead of `-`

**Comment by ggerganov** [comment]
Perfect - the tokenizer test don't work because they need new vocabs:

```bash
 $ ▶ ./bin/test-tokenizer-0 ../models/ggml-vocab-llama.bin 
main : reading vocab from: '../models/ggml-vocab-llama.bin'
gguf_init_from_file: invalid magic number 67676a74
Segmentation fault (core dumped)
```

We need to generate new vocab files.

**Comment by klosax** [comment]
Maybe better to use a model file instead of using a separate vocab? And have information that the tests only works with the original LLaMA models.

**Comment by klosax** [comment]
Later we will need tokenizer tests for other models like mpt falcon gptneox..

**Comment by ggerganov** [comment]
We need small vocab files inside the repository in order to run the CI

**Comment by klosax** [comment]
Ok. The vocab files are model files without the tensors, right?

**Comment by klosax** [comment]
Should not have tensor_count = 0 and include tensor info? Maybe skip tensor info?

**Comment by monatis** [comment]
Setting tensor count to 0 perfectly works without breaking file format --I previously used it for testing.

We can define a flag in conversion script `metadata_only = False` and setting it to `True` can output metadata-only file without tensors for that model.

**Comment by ggerganov** [comment]
There is already `--vocab-only` arg in `convery.py`:

https://github.com/ggerganov/llama.cpp/blob/5b94b14d5dd59e14f2ab443a79413c445726c98f/convert.py#L1305

I consider the other `convert` scripts that we currently use just temporary until we update `convert.py` to work with `gguf.py`, so not sure it's worth adding more features to them

**Comment by klosax** [comment]
> Edit: this is regarding the LLaMA models. For other models, we should likely have separate convert scripts

I that case i think the lazy tensor reader from convert.py should be separated so they can be used by the other scripts.

**Comment by klosax** [comment]
LLaMA v1 13b main and perplexity works

**Comment by klosax** [comment]
LLaMA v1 33b works in main and perplexity

**Comment by klosax** [comment]
OpenLLaMA v2 3b and 7b works main perplexity

**Comment by klosax** [comment]
LLaMA v1 7b also works.

**Comment by ggerganov** [comment]
Weird that 70B does not work - I can't test as I don't have HF 70B. Just the original one

Will merge like this and fix the convert script. When ready, will test 70B and try to resolve

**Comment by Green-Sky** [comment]
> Weird that 70B does not work - I can't test as I don't have HF 70B. Just the original one
> 
> Will merge like this and fix the convert script. When ready, will test 70B and try to resolve

iirc there was a fix on master after gguf branched. (cant find it rn so might be bad memory)

**Comment by klosax** [comment]
> iirc there was a fix on master after gguf branched.

If you was thinking of this https://github.com/ggerganov/llama.cpp/pull/2427 it is already included.
I cant fully test the 70b model on my system.

**Comment by Green-Sky** [comment]
> If you was thinking of this https://github.com/ggerganov/llama.cpp/pull/2427 it is already included.
I cant fully test the 70b model on my system.

hm yea, i think i meant that. never mind then.

**Comment by goerch** [comment]
Just tried to the branch. Converting with

```
python.exe convert-llama-h5-to-gguf.py models\llama-7B-hf 1
```

looks fine:

```
output.weight, shape 2, torch.float16 --> float16
gguf: model successfully exported to 'models\llama-7B-hf/ggml-model-f16.gguf'
```

The resulting file `gguf` file is around 13 GB big. When starting inference I get

```
main: build = 1168 (d864596)
main: seed  = 1692248927
gguf_init_from_file: invalid magic number 00000000
```

This happens with clang-cl on Windows.

**Comment by klosax** [comment]
> gguf_init_from_file: invalid magic number 00000000

I got this when accidentally tried to open a directory instead of a model file.

**Comment by monatis** [comment]
> I got this when accidentally tried to open a directory instead of a model 

Me too, and reproduceable.

```sh
$ ./main -m ~                                                       

main: build = 1157 (5ec1893)                                                                                            
main: seed  = 1692257530                                                                                                
gguf_init_from_file: invalid magic number 00000000                                                                      
Segmentation fault                                                                                                      
```

**Comment by klosax** [comment]
> gguf: model successfully exported to 'models\llama-7B-hf/ggml-model-f16.gguf'

There is both backslashes and slashes in this path, dont know if you are using this to load?

**Comment by klosax** [comment]
I get the error when trying to open a directory instead of a model file. On ubuntu.

**Comment by goerch** [comment]
My mistake was trying to open an non existing file.

**Comment by monatis** [comment]
> does the wrong magic of 0000000 occur for directories, or does it occur with a converted F16 model?

To my end, I couldn't reproduce with a GGUF file, but passing a directory path leads to wrong magic on WSL, regular Windows and regular Ubuntu.

**Comment by monatis** [comment]
Maybe we can make that error message a little bit verbose with "Are you sure that this path holds a GGUF file?"

**Comment by klosax** [comment]
Better " Can not open an directory. Should be a .gguf model file."

**Comment by ggerganov** [comment]
@goerch 

I'm running the `test-tokenizer-0` test using the new F16 GGUF model, and it is currently failing.
We made the following change recently:

from:
```c++
static uint8_t llama_byte_to_char(const llama_vocab & vocab, uint8_t byte) {
    if (llama_vocab_type(vocab) == "spm") {
        return byte + 3;
    }
```

to:
```c++
        return byte - 3;
```

This was needed because the new-line wasn't printing correctly.
However, with this change, the `test-tokenizer-0` is now failing.

If I revert it, it no longer fails, but the de-tokenized strings do not match the original strings.
For example:

```
main : ' this is 🦙.cpp' tokenized to '  this is Ꙡ.cpp'
```

Will be looking into this now - any help would be appreciated

**Comment by ggerganov** [comment]
I think I fixed it:

https://github.com/ggerganov/llama.cpp/pull/2635/commits/7b6ae8904178cb8f269719545a0396e5f726ff22

We were incorrectly using `llama_byte_to_char()` when tokenizing a string.
To fix it, I added `llama_char_to_byte()` and use it instead

File: llama.cpp [hunk 1]
```diff
@@ -447,726 +936,510 @@ struct llama_context {
...
+            /*.ctx      = */ &ctx_meta,
+        };
 
-        // LLaMAv2
-        // TODO: read from header
-        hparams.n_head_kv = hparams.n_head;
-    }
-    void read_vocab() {
-        vocab.id_to_token.resize(hparams.n_vocab);
+        ctx_gguf = gguf_init_from_file(fname.c_str(), params);
```
Comment (Cebtenzzre) [inline llama.cpp:1043]: There's a missing check for `ctx_gguf == NULL` here, which is why "invalid magic number" leads to a segfault.

**Comment by Cebtenzzre** [comment]
And checking ferror(file) if gguf_fread_el returns false here should make it more obvious if the user passes a directory instead of a file:
https://github.com/ggerganov/llama.cpp/blob/795ec7070cba4356112836bd0de3a718cb04e8bf/ggml.c#L18717-L18721

**Comment by klosax** [comment]
I think it is better if you review here instead https://github.com/ggerganov/llama.cpp/pull/2398

**Pull request merged**